File: Assets/Scripts/Player/PlayerStamina.cs
using UnityEngine;
using Mirror;
using System;

namespace ZombieSurvival.Player
{
    public class PlayerStamina : NetworkBehaviour
    {
        [Header("Stamina Settings")]
        [SerializeField] private float maxStamina = 100f;
        [SerializeField] private float staminaDrainRate = 15f;
        [SerializeField] private float staminaRegenRate = 10f;
        [SerializeField] private float regenDelay = 1.5f;
        [SerializeField] private float exhaustionThreshold = 0f;
        [SerializeField] private float exhaustionRecovery = 20f;

        [SyncVar(hook = nameof(OnStaminaChanged))]
        private float currentStamina;

        private float lastUseTime;
        private bool isExhausted;

        public float CurrentStamina => currentStamina;
        public float MaxStamina => maxStamina;
        public float StaminaPercent => currentStamina / maxStamina;
        public bool IsExhausted => isExhausted;

        public event Action<float, float> OnStaminaUpdated; // current, max

        public override void OnStartServer()
        {
            base.OnStartServer();
            currentStamina = maxStamina;
        }

        private void Update()
        {
            if (!isServer) return;

            if (Time.time - lastUseTime >= regenDelay)
            {
                if (currentStamina < maxStamina)
                {
                    currentStamina = Mathf.Min(currentStamina + staminaRegenRate * Time.deltaTime, maxStamina);
                }

                if (isExhausted && currentStamina >= exhaustionRecovery)
                {
                    isExhausted = false;
                }
            }
        }

        [Server]
        public void UseStamina(float amount)
        {
            if (isExhausted) return;

            currentStamina = Mathf.Max(0, currentStamina - amount * staminaDrainRate);
            lastUseTime = Time.time;

            if (currentStamina <= exhaustionThreshold)
            {
                isExhausted = true;
                RpcOnExhausted();
            }
        }

        [Server]
        public void RestoreStamina(float amount)
        {
            currentStamina = Mathf.Min(currentStamina + amount, maxStamina);
        }

        [ClientRpc]
        private void RpcOnExhausted()
        {
            AudioManager.AudioManager.Instance?.PlaySFX("exhausted");
        }

        private void OnStaminaChanged(float oldStamina, float newStamina)
        {
            OnStaminaUpdated?.Invoke(newStamina, maxStamina);
        }
    }
}

File: js/player/PlayerStamina.js
// PlayerStamina.js - Stamina system
class PlayerStamina {
    constructor() {
        this.max = 100;
        this.current = 100;
        this.drainRate = 15;
        this.regenRate = 12;
        this.regenDelay = 1.5;
        this.lastUseTime = 0;
        this.isExhausted = false;
        this.exhaustionRecovery = 25;
    }

    get percent() { return this.current / this.max; }

    use(amount) {
        if (this.isExhausted) return false;
        this.current = Math.max(0, this.current - amount);
        this.lastUseTime = performance.now() / 1000;

        if (this.current <= 0) {
            this.isExhausted = true;
        }
        return true;
    }

    update(deltaTime) {
        const now = performance.now() / 1000;

        if (now - this.lastUseTime >= this.regenDelay) {
            if (this.current < this.max) {
                this.current = Math.min(this.current + this.regenRate * deltaTime, this.max);
            }
            if (this.isExhausted && this.current >= this.exhaustionRecovery) {
                this.isExhausted = false;
            }
        }
    }

    reset() {
        this.current = this.max;
        this.isExhausted = false;
    }
}
